### PR TITLE
[SYCL-MLIR]: Loop Versioning - introduce LoopVersionCondition

### PR DESCRIPTION
**Pull request by etiotto**
Create the `LoopVersionCondition` class to abstract the loop versioning condition used to version a loop. 
Create the `LoopTools` class to collect tools useful for loop transformations.
Adapt `LICM` to create the  `LoopVersionCondition` class and use `LoopTools`. Loop versioning is now requested via: 
```
std::unique_ptr<LoopVersionCondition> condition =
   VersionConditionBuilder(loop, accessorPairs).createCondition();
loopTools.versionLoop(loop, *condition);
```

Now there is a single API for versioning SCF or Affine loops. The consumer uses a  `VersionConditionBuilder` to create an appropriate condition for the kind of loop being versioned and `LoopTools` take care of the rest.

### CODE REVIEW

File: polygeist/include/mlir/Dialect/Polygeist/Utils/TransformUtils.h [hunk 1]
```diff
@@ -53,19 +53,47 @@ class IfThenElseBodyInterface {
...
+  using SCFCondition = Value;
+
+  struct AffineCondition {
+    IntegerSet ifCondSet;
+    SmallVectorImpl<Value> &setOperands;
+  };
+
+  /// Create a loop versioning condition for SCF loops.
+  LoopVersionCondition(SCFCondition scfCond)
+      : versionCondition({&scfCond, nullptr}) {}
```
Comment (whitneywhtsang) [inline polygeist/include/mlir/Dialect/Polygeist/Utils/TransformUtils.h:68]: would scfCond be out of scope after the constructor?

File: polygeist/lib/Dialect/Polygeist/Transforms/LICM.cpp [hunk 2]
```diff
@@ -770,11 +794,9 @@ static size_t moveLoopInvariantCode(LoopLikeOpInterface loop,
...
     if (!accessorPairs.empty()) {
       OpBuilder builder(loop);
-      // TODO: find a way to create a versioning condition that works for
-      // creating either SCF loops or affine loops.
-      Value condition = VersionConditionBuilder(accessorPairs)
-                            .createConditionForSCFLoop(builder, loop.getLoc());
-      LoopVersionBuilder::create(loop)->versionLoop(condition);
+      std::unique_ptr<LoopVersionCondition> condition =
+          VersionConditionBuilder(loop, accessorPairs).createCondition();
+      LoopVersionBuilder::create(loop)->versionLoop(*condition);
```
[the remark this answers is not in the document]
Comment (etiotto) [reply polygeist/lib/Dialect/Polygeist/Transforms/LICM.cpp]: `VersionConditionBuilder::createCondition` needs to create a new `LoopVersionCondition` object and return a pointer to it. 
It it returns a C pointer the caller would then have to delete it while with unique_ptr there is no need to delete the pointer, preempting the risk of a memory leak.

File: polygeist/include/mlir/Dialect/Polygeist/Utils/TransformUtils.h (hunk 1 above)
Comment (etiotto) [reply polygeist/include/mlir/Dialect/Polygeist/Utils/TransformUtils.h]: This has changed to use an Optional

File: polygeist/include/mlir/Dialect/Polygeist/Utils/TransformUtils.h [hunk 3]
```diff
@@ -53,19 +53,50 @@ class IfThenElseBodyInterface {
...
+    return *versionCondition.scfCond;
+  }
+  AffineCondition getAffineCondition() const {
+    assert(versionCondition.affineCond.has_value() &&
+           "expecting valid affine condition");
+    return *versionCondition.affineCond;
+  }
+
+private:
+  struct VersionCondition {
```
Comment (victor-eds) [inline polygeist/include/mlir/Dialect/Polygeist/Utils/TransformUtils.h:85]: We could use `std::variant` instead.

Comment (etiotto) [reply polygeist/include/mlir/Dialect/Polygeist/Utils/TransformUtils.h:85]: I was initially thinking that some consumer could create a versioning condition holding both a valid  `SCFCondition` and an `AffineCondition` but I think that is unlikely and probably overkill. I'll make the change to use a std::variant.

**Review by victor-eds — APPROVED** [review]

**Comment by etiotto** [comment]
The failure in E2E test is not a regression caused by this PR so merging.

**Pull request merged**